src/view/wlr-surface-node.cpp

```diff
@@ -193,13 +193,19 @@ wf::touch_interaction_t& wf::scene::wlr_surface_node_t::touch_interaction()
     return *this->tch_interaction;
 }
 
-void wf::scene::wlr_surface_node_t::send_frame_done()
+void wf::scene::wlr_surface_node_t::send_frame_done(bool delay_until_vblank)
 {
-    if (surface)
+    if (surface && (!delay_until_vblank || visibility.empty()))
     {
         timespec now;
         clock_gettime(CLOCK_MONOTONIC, &now);
         wlr_surface_send_frame_done(surface, &now);
+    } else if (surface)
```

Comment:
since you're checking `surface != NULL` in both cases, maybe just wrap these two conditions in `if (surface)`.